Site Assembler - Add a category to explore All patterns

miksansegundo opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #76451 #76580

## Proposed Changes


https://user-images.githubusercontent.com/1881481/236999478-dd421c9a-2df9-4aa0-809b-c7e7abb5822f.mov



* Add a category at the top of the list to explore `All` patterns in one list
* Shuffle the patterns in the category `All` to get better stats of popular patterns
* Use the feature flag `pattern-assembler/all-patterns-category` (enabled on calypso, horizon and dev envs)
* Use a real pattern category (the first in the `pattern.categories`) instead of the category `All` as the pattern name on the list of selected patterns
* All Tracks events work as usual, we use the category slug `all` for patterns selected from the `All` category 

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

**This PR depends on #76580 because the category `All` is registered from the ETK plugin.**

- Follow the testing instructions on #76580 to install the ETK plugin
- Create a new site with any plan
- Continue until the Design Picker
- Scroll down and click `Start designing`
- Click `Sections` > `Add patterns` > `All`
- Scroll the list to verify that all patterns are there
- Verify that this change doesn't impact the performance
- Add patterns from the `All` category and verify that on the list of selected patterns the name used is not `All` but a real pattern category
- Verify that `all` appears in the Tracks event properties `pattern_category` and `pattern_categories`

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [x] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [x] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [x] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [x] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [x] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

miksansegundo committed:
Add a category to explore All patterns in one list

miksansegundo committed:
Use first category as pattern name in main list

arthur791004 commented:
hmmm...the categories are from the endpoint so we cannot see the `ALL` categories in the list 😂

miksansegundo commented on client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/hooks/use-category-all.ts:
```diff
@@ -0,0 +1,27 @@
+				categories: {
+					...pattern.categories,
+					all: {
+						slug: 'all',
```
@arthur791004, if we agree to use the slug `featured` for the category `All`, we'll have to update this PR to use it.

miksansegundo committed:
Use the category slug featured as a workaround to put the category all in the first position

miksansegundo committed:
Add featured category slug for category all

autumnfjeld commented:
@miksansegundo  @alaczek  Just curious: should "All" be the first category?    (Maybe this was already discussed and I missed it.)

autumnfjeld commented:
> @autumnfjeld let me know if this strategy to track the usage of the category All only with the event calypso_signup_pattern_assembler_pattern_select_click works.

>  I decided to use the first category of the pattern because that could give us an idea of the most popular categories picked from the category All.

What is the "first category of the pattern"? Is that some special designation?  

The tracking for the category All should be treated just like any other category.    Because one part of the analysis is "What did the user click".  So if the user clicked All, we should record All.  

And because each pattern can belong to many categories, it seems strange to arbitrarily pick "some other category" than the one the user actually clicked on.

arthur791004 commented on client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/index.tsx:
```diff
@@ -315,8 +315,8 @@ const PatternAssembler = ( {
+			// or the first category because it's used in tracks and as pattern name in the list
 			const [ firstCategory ] = Object.keys( selectedPattern.categories );
 			selectedPattern.category = categories.find(
 				( { name } ) => name === ( selectedCategory || firstCategory )
```
I guess we can simply check the `selectedCategory` here and then we don't need to find the real category later

```js
selectedPattern.category = categories.find(
  ( { name } ) => {
    if ( selectedCategory === CATEGORY_ALL_SLUG ) {
      return name === firstCategory;
    }

    return name === ( selectedCategory || firstCategory );
  }
)
```

arthur791004 commented on client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/index.tsx:
```diff
@@ -330,6 +330,10 @@ const PatternAssembler = ( {
 			} );
```
I think we have to use `all` instead of `featured` for tracks events. WDYT?

miksansegundo committed:
Fix tracks events

miksansegundo committed:
Fix function name

miksansegundo commented:
I updated the description of the issue @autumnfjeld, and updated the PR: 

- Add a category **at the top of the list** to explore `All` patterns in one list
- Use **a real pattern category (the first in the pattern.categories)** instead of the category All as the pattern name on the list of selected patterns
- **All Tracks events work as usual**, we use the category slug `all` for patterns selected from the All category

<img width="771" alt="Screenshot 2566-05-09 at 11 58 35" src="https://user-images.githubusercontent.com/1881481/236999947-23bb169e-88ea-4c53-b500-b69d0d57d167.png">

arthur791004 commented on client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/index.tsx:
```diff
@@ -316,11 +316,23 @@ const PatternAssembler = ( {
+				if ( selectedCategory === CATEGORY_ALL_SLUG ) {
+					return name === firstCategory;
+				}
+				return name === ( selectedCategory || firstCategory );
```
I just found out there is an issue here before. If the first category is in front of the selected category, we will use the first category instead of the selected one.

miksansegundo replied:
Just to confirm, it is now solved in this version, right?

Could you help me understand the issue? I don't get what you mean by "If the first category is in front of the selected category".

arthur791004 replied:
For example, 

```js
const categories = [ 'text', 'blog' ];
const selectedCategory = 'blog';
const selectedPattern = {
  categories: {
    'text': ...,
    'blog': ...,
  }
}

const [ firstCategory ] = Object.keys( selectedPattern.categories );
// firstCategory: text
```

Next, when we loop through the `categories`, we will get `text` instead of the blog because the `firstCategory` will be matched at the beginning.

miksansegundo replied:
Do you mean that will be matched in this line `name === ( selectedCategory || firstCategory )`?

arthur791004 replied:
Ooops...I found I misread something...you can ignore what I said 🤦‍♂️

arthur791004 reviewed (approved):
👍
